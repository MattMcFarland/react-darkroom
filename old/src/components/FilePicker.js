import React from 'react';

export default class FilePicker extends React.Component {
  render() {

    let {onChange, hasFile} = this.props;
    let btnStyle = {
      display: hasFile ? 'none' : 'block',
      width: '100%',
      background: 'inherit',
      border: 'inherit',
      color: '#bbb'
    };
    let iconStyle = {
      width: '100%',
      display: 'block',
      fontSize: '240px'
    };
    let textStyle = {
      fontSize: '34px'
    };

    return (
      <button style={btnStyle} onClick={() => this.refs.fileInput.click()} type="button" className="filepicker">
        <span style={iconStyle} className="icon icon-image"/>
        <span style={textStyle}>Select image</span>
        <input ref="fileInput" style={{display: 'none'}} type="file" onChange={onChange}/>
      </button>
    );
  }
}
