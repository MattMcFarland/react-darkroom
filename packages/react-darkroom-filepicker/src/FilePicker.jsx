import React from 'react';
import PropTypes from 'prop-types';
import injectSheet from 'react-jss';
import {
  compose,
  pure,
  setDisplayName,
  withHandlers,
  defaultProps,
} from 'recompose';

/* global window */

let fileInput = null;

const readFile = (file, done) => {
  const reader = new window.FileReader();
  reader.onload = e => done(e.target.result);
  reader.readAsDataURL(file);
};

export const FilePicker = ({
  classes,
  className,
  children,
  getFileData,
  triggerInputChange,
}) => (
  /* eslint jsx-a11y/no-static-element-interactions: 0 */
  <span aria-label="select file">
    <span
      onClick={triggerInputChange}
      className={className || classes.filePicker}
    >
      {children}
    </span>
    <input
      onChange={getFileData}
      type="file"
      style={{ display: 'none' }}
      ref={(input) => { fileInput = input; }}
    />
  </span>
);

FilePicker.propTypes = {
  /** Us
   *  <button></button>e a child or children to design the look of the component, 
   * wrapped in a button tag. */
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.element, PropTypes.string]).isRequired,
  /** Leave empty to use default styling, which by default inherits it's parents attributes */
  className: PropTypes.string,
  /** @ignore */
  classes: PropTypes.shape({
    filePicker: PropTypes.string,
  }).isRequired,
  /* eslint react/no-unused-prop-types: 0 */
  /** Event handler for when the file data has loaded, which async occurs after user selects file */
  onFileDataLoaded: PropTypes.func,
  /** Event handler for when an error occurs while loading the file from the user's system */
  onFileDataError: PropTypes.func,
  /** @ignore */
  getFileData: PropTypes.func.isRequired,
  /** @ignore */
  triggerInputChange: PropTypes.func.isRequired,
};

FilePicker.defaultProps = {
  onFileDataError: (e) => { throw e; },
  onFileDataLoaded: () => {},
  className: null,
};

const styles = {
  filePicker: {
    border: 'inherit',
    background: 'inherit',
    textAlign: 'left',
    fontSize: 'inherit',
    color: 'inherit',
    display: 'inherit',
  },
};

export default compose(
  pure,
  defaultProps({
    fileDataError: (e) => { throw e; },
    fileDataLoaded: () => {},
  }),
  withHandlers({
    getFileData: props => (event) => {
      event.persist();
      try {
        readFile(event.target.files[0], (data) => {
          const {
            lastModified,
            name,
            size,
            type,
          } = event.target.files[0];
          props.fileDataLoaded({ data, meta: { lastModified, name, size, type } });
        });
      } catch (e) {
        props.fileDataError(e);
      }
    },
    triggerInputChange: () => () => {
      fileInput.click();
    },
  }),
  injectSheet(styles),
  setDisplayName('FilePicker'),
)(FilePicker);
