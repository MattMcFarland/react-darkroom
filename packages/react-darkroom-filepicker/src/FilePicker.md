### FilePicker Basics

FilePicker is used primarily to handle the operation of loading a file into memory.  It allows
full customization, and by default it looks intentionally plain.

    <FilePicker fileDataLoaded={ ({meta}) => {
      console.log('container 1')
      setState({meta})
    }}>
      <span style={{ color: 'red' }}>Click me...</span>
      <pre>
        {state.meta && JSON.stringify(state, null, 2)}
      </pre>
    </FilePicker>

### Customize appearance

You can customize the filepicker's appearance by either providing a your own `className` prop to it, or by providing your own children.  

#### Custom class 

    <FilePicker fileDataLoaded={ ({meta}) =>       
      setState({meta2: meta})
    }>
      <span style={{ color: 'red' }}>Click me...</span>
      <pre>
        {state.meta2 && JSON.stringify(state, null, 2)}
      </pre>
    </FilePicker>