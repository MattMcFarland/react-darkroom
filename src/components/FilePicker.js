import React from 'react';

export default class FilePicker extends React.Component {

  render () {

    let {onChange, hasFile} = this.props,
      btnStyle = {
        display: hasFile ? 'none': 'block',
        width: '100%',
        background: 'inherit',
        border: 'inherit',
        color: '#bbb'
      },
      iconStyle = {
        width: '100%',
        display: 'block',
        fontSize: '240px'
      },
      textStyle = {
        fontSize: '34px'
      },
      onClick = () => {
        this.refs.fileInput.click();
      };
    return (
      <button style={btnStyle} onClick={onClick} type="button" className="filepicker">
        <span style={iconStyle} className="icon icon-image"/>
        <span style={textStyle}>Select image</span>
        <input ref="fileInput" style={{display: 'none'}} type="file" onChange={onChange}/>
      </button>
    )
  }
}
  /*
  onChange,
  hasFile,
  onClick:  => {

  },
  }) => (

);
*/
