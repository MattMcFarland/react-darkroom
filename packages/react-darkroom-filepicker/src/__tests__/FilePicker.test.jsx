import React from 'react';

import { shallow, mount, render } from 'enzyme';
import * as FilePickerModule from '../FilePicker';

const FilePicker = FilePickerModule.default;
const FilePickerClass = FilePickerModule.FilePicker;

const { describe, test, expect } = global;

const FilePickerNode = (
  <FilePicker>
    <button>test</button>
  </FilePicker>);

const createFilepicker = () => mount(FilePickerNode).find(FilePickerClass);


describe('FilePicker', () => {
  describe('Rendering', () => {
    test('should render an input field as a child', () => {
      expect(createFilepicker().find('input').length).toEqual(1);
    });
    test('multiple filepickers must contain their own UUIDs to have unique refs', () => {
      const uuidHash =
        [createFilepicker(), createFilepicker(), createFilepicker(), createFilepicker()]
          .map(fp => fp.props().uuid)
          .reduce((hash, uuid) => Object.assign(hash, { [uuid]: uuid }), {});
      expect(Object.keys(uuidHash).length).toEqual(4);
    });
  });
  describe('onClick', () => {
    test('triggers the call of click on the hidden input tag', () => {
      const wrapper = createFilepicker();
      const input = wrapper.find('input').getNode();

      input.click = jest.fn();
      wrapper.simulate('click');
      expect(input.click).toBeCalled();
    });
  });
  describe('Snapshot', () => {
    test('matches snapshot', () => {
      const wrapper = createFilepicker();
      expect(wrapper.render()).toMatchSnapshot();
      wrapper.simulate('click');
      expect(wrapper.render()).toMatchSnapshot();
    });
  });
});
