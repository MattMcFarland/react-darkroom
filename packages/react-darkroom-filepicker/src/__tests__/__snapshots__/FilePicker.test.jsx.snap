// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilePicker Snapshot matches snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "aria-label": "select file",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "filePicker-0-0",
            },
            "children": Array [
              Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "data": "test",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "button",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
              },
            ],
            "name": "span",
            "next": Object {
              "attribs": Object {
                "style": "display: none;",
                "type": "file",
              },
              "children": Array [],
              "name": "input",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
          },
          Object {
            "attribs": Object {
              "style": "display: none;",
              "type": "file",
            },
            "children": Array [],
            "name": "input",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "filePicker-0-0",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "data": "test",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "button",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                },
              ],
              "name": "span",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
            },
            "type": "tag",
          },
        ],
        "name": "span",
        "next": null,
        "parent": null,
        "prev": null,
        "root": [Circular],
        "type": "tag",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xmlMode": false,
  },
}
`;

exports[`FilePicker Snapshot matches snapshot 2`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "aria-label": "select file",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "filePicker-0-0",
            },
            "children": Array [
              Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "data": "test",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "button",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
              },
            ],
            "name": "span",
            "next": Object {
              "attribs": Object {
                "style": "display: none;",
                "type": "file",
              },
              "children": Array [],
              "name": "input",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
          },
          Object {
            "attribs": Object {
              "style": "display: none;",
              "type": "file",
            },
            "children": Array [],
            "name": "input",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "filePicker-0-0",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "data": "test",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "button",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                },
              ],
              "name": "span",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
            },
            "type": "tag",
          },
        ],
        "name": "span",
        "next": null,
        "parent": null,
        "prev": null,
        "root": [Circular],
        "type": "tag",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xmlMode": false,
  },
}
`;
